src/App.tsx

```diff
@@ -15,36 +16,124 @@ export const goodsFromServer = [
+        case SortType.SORT_BY_ALPHABET:
+          return sortByAlphabet(isReversed, good1, good2);
+
+        case SortType.SORT_BY_LENGTH:
+          if (good1.length === good2.length) {
+            return sortByAlphabet(isReversed, good1, good2);
+          }
+
+          return isReversed
+            ? good2.length - good1.length
+            : good1.length - good2.length;
+
+        default:
+          return 0;
+      }
+    });
+  }
+
+  return updateGoods;
+}
+
 export const App: React.FC = () => {
+  const [sortBy, setSortBy] = useState<SortType | null>(null);
+  const [isReversed, setIsReversed] = useState(false);
+
+  const sortedGoods = getSortedGoods(goodsFromServer, { sortBy, isReversed });
+
+  const toggleReverse = () => setIsReversed(prevIsReversed => !prevIsReversed);
+
+  const reset = () => {
+    setSortBy(null);
+    setIsReversed(false);
+  };
+
+  const isActivButton = (
+    stateValue: SortType | null | boolean,
+    activeCondition: SortType | boolean,
+  ) => (activeCondition !== stateValue ? 'is-light' : '');
+
   return (
     <div className="section content">
       <div className="buttons">
-        <button type="button" className="button is-info is-light">
+        <button
+          type="button"
+          onClick={() => setSortBy(SortType.SORT_BY_ALPHABET)}
+          className={classNames(
+            'button',
+            'is-info',
+            isActivButton(sortBy, SortType.SORT_BY_ALPHABET),
+          )}
+        >
           Sort alphabetically
         </button>
 
-        <button type="button" className="button is-success is-light">
+        <button
+          type="button"
+          onClick={() => setSortBy(SortType.SORT_BY_LENGTH)}
+          className={classNames(
+            'button',
+            'is-success',
+            isActivButton(sortBy, SortType.SORT_BY_LENGTH),
+          )}
+        >
           Sort by length
         </button>
 
-        <button type="button" className="button is-warning is-light">
+        <button
+          type="button"
+          onClick={toggleReverse}
+          className={classNames(
+            'button',
+            'is-warning',
+            isActivButton(isReversed, true),
+          )}
+        >
           Reverse
         </button>
 
-        <button type="button" className="button is-danger is-light">
-          Reset
-        </button>
+        {(sortBy || isReversed) && (
+          <button
+            type="button"
+            onClick={reset}
+            className="button is-danger is-light"
+          >
+            Reset
+          </button>
+        )}
       </div>
 
       <ul>
-        <ul>
-          <li data-cy="Good">Dumplings</li>
-          <li data-cy="Good">Carrot</li>
-          <li data-cy="Good">Eggs</li>
-          <li data-cy="Good">Ice cream</li>
-          <li data-cy="Good">Apple</li>
-          <li data-cy="Good">...</li>
-        </ul>
+        {sortedGoods.map(good => (
+          <li data-cy="Good" key={good}>
```

Comment:
Using the `good` string as a key in the list may lead to issues if the list contains duplicate items. Consider using a unique ID or a combination of the item's content and index to ensure the key is unique. [CODE KNOWLEDGE]